src/lib/router.ts

```diff
@@ -100,31 +111,58 @@ export function updateProjectRoute(
   }
 }
 
-export const push = (newRoute: Route): void => {
+const historyExecutor = mutexExecutor.create();
+
+async function setHistory(history: Route[]): Promise<void> {
+  showLoadingBar.set(true);
+  const newRoute = history.slice(-1)[0];
+
+  // TODO: is there a way to somehow abort the previous fetches if we
+  // receive new navigation requests?
+  const loadedRoute = await historyExecutor.run(async () => {
+    if (newRoute.resource === "seeds") {
+      return await loadSeedRoute(newRoute.params);
+    } else if (newRoute.resource === "home") {
+      return await loadHomeRoute();
+    } else {
+      return newRoute;
+    }
+  });
+
+  // Only let the last request through.
+  if (loadedRoute === undefined) {
+    return;
+  }
+
+  historyStore.set(history);
+  activeRouteStore.set(loadedRoute);
+  showLoadingBar.set(false);
+}
+
+export async function push(newRoute: Route): Promise<void> {
   const history = get(historyStore);
 
-  // Limit history to a maximum of 10 steps. We shouldn't be doing more than
-  // one subsequent pop() anyway.
-  historyStore.set([...history, newRoute].slice(-10));
+  // Limit history to a maximum of 100 steps.
+  await setHistory([...history, newRoute].slice(-100));
 
   const path = import.meta.env.VITE_HASH_ROUTING
     ? "#" + routeToPath(newRoute)
     : routeToPath(newRoute);
 
   window.history.pushState(newRoute, documentTitle, path);
-};
+}
 
-export const pop = (): void => {
+export async function pop(): Promise<void> {
   const history = get(historyStore);
   const newRoute = history.pop();
   if (newRoute) {
-    historyStore.set(history);
+    await setHistory(history);
     window.history.back();
```

Comment:
Is this still an issue after all the refactoring we did?

Reply:
Yes, this is still an issue, albeit a smaller one. Like in the comment above, let’s assume the user clicks a link. This results in a call to `push()`, which loads the new route, say, the seed page. While the loading is in progress the user clicks on a different link to, say, the home. Now the home loading is initiated and the loading of the seed page is aborted. This means in the `push(seedRoute)` call `await mutexLoad(seedRoute)` returns undefined and the seed route is pushed to the history and the path is changed. Then, the home route finishes loading, sets the active route, which is rendered, and in `push(homeRoute)` `await mutexLoad(homeRoute)` returns and the home route is pushed to the history and the path is changed.

Now, when the user clicks back, they will land on the seed route, because that was pushed to the history earlier although the navigation was aborted. In addition the user will also have seen the path change without the path being rendered when they aborted the navigation by click on the home route.

You can argue that these issues are not super critical. But I think it’s worth tackling this problem to provide a  solid and correct foundation for the concurrent behavior that will survive refactorings. This is also a good example how to write better code dealing with concurrency.

The way to think about this is that navigation (with push or replace) is an atomic transaction with the following properties
1. Only one transaction can be active at a given time
2. When a new transaction is started, the previous one is aborted
3. When a transaction is aborted none of it’s side effects should be visible to the outside (atomic)

What the current code is violating is the last property. With transactions it is a good idea to put all of the code into one function so it is easy to verify that the transaction is atomic. This means we should move the `history.pushState()` side effect into `mutexLoad`. Here’s an example how to do this.

```typescript
async function navigate(
  action: "push" | "replace",
  newRoute: Route,
): Promise<void> {
  isLoading.set(true);

  const loadedRoute = await loadExecutor.run(async () => {
    return loadRoute(newRoute);
  });

  // Only let the last request through.
  if (loadedRoute === undefined) {
    return;
  }

  activeRouteStore.set(loadedRoute);
  isLoading.set(false);

  const path = import.meta.env.VITE_HASH_ROUTING
    ? "#" + routeToPath(newRoute)
    : routeToPath(newRoute);

  if (action === "push") {
    window.history.pushState(newRoute, DOCUMENT_TITLE, path);
  } else if (action === "replace") {
    window.history.replaceState(newRoute, DOCUMENT_TITLE, path);
  }
}

export async function push(newRoute: Route): Promise<void> {
  await navigate("push", newRoute);
}

export async function replace(newRoute: Route): Promise<void> {
  await navigate("replace", newRoute);
}
```
(I’ve also renamed `mutexLoad` to `navigate`)